Carthage XCF support and OpenSSL 1.1.1k

julepka opened the pull request:
Last Friday [Carthage released an important update](https://github.com/Carthage/Carthage/releases/tag/0.38.0): now, it supports binary dependencies with XCF. It means that Themis for Carthage can use OpenSSL XCF binary, and Themis itself will be build as an XCF. 🎉🎉🎉

What's new
* Themis Carthage moved to XCF
* Themis Carthage uses OpenSSL 1.1.1k (latest)
* No extra steps with `Themis.xcodeproj` required to build `themis.xcframework` with a script.

## Checklist

- [x] Change is covered by automated tests 
- [x] The [coding guidelines] are followed
- [x] Example projects and code samples are up-to-date (in case of API changes) - _We are going to update them after the release, cuz... no release - no examples :)_
- [x] Changelog is updated (in case of notable or breaking changes)

[coding guidelines]: https://github.com/cossacklabs/themis/blob/master/CONTRIBUTING.md

vixentael commented on Themis.xcodeproj/project.pbxproj:
```diff
@@ -3,10 +3,20 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		6DB8748A26497B8C0022A1F6 /* openssl.xcframework in Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; };
+		6DB8748E26497B950022A1F6 /* openssl.xcframework in Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; };
+		6DB8749326497BEF0022A1F6 /* openssl.xcframework in Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; };
+		6DB8749426497C1E0022A1F6 /* openssl.xcframework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		6DB8749726497C240022A1F6 /* openssl.xcframework in Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; };
+		6DB8749A26497C2B0022A1F6 /* openssl.xcframework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		6DB8749D26497C300022A1F6 /* openssl.xcframework in Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; };
+		6DB874A026497C350022A1F6 /* openssl.xcframework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		6DB874A326497C390022A1F6 /* openssl.xcframework in Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; };
+		6DB874A626497C3E0022A1F6 /* openssl.xcframework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 6DB8748926497B8B0022A1F6 /* openssl.xcframework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
```
that's a lot of frameworks :D

vixentael commented on Themis.xcodeproj/project.pbxproj:
```diff
@@ -1567,6 +1569,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				CURRENT_PROJECT_VERSION = 8;
 				DEFINES_MODULE = YES;
+				DEVELOPMENT_ASSET_PATHS = "";
```
👀

vixentael commented on CHANGELOG.md:
```diff
@@ -4,6 +4,21 @@
+
+  - `themis` for Carthage now loads OpenSSL as xcframework, and Carthege builds `themis` as scframework as well.
+  - Updated OpenSSL to the latest 1.1.1k for Carthage
+  - `create_xcframeworks.sh` script (used for SPM) no longer requires `Themis.xcodeproj` modifications because OpenSSL is now added as xcframework.
```
`create_xcframeworks.sh` was this script changed? I don't see it in PR

julepka replied:
Themis.xcodeproj contained OpenSSL linked with Carthage. It was an older version of OpenSSL. SMP supported the newer OpenSSL version, so you had to go to the Themis.xcodeproj and add it before building the XCF. Now we can forget about all these manipulations because Carthage is adding the same XCF as SPM.

vixentael reviewed (changes requested):
see above

julepka commented:
> Also I read on Carthage release page, that building process requires running Carthage twice
> 
> ```
> Support for making an xcframework archive will be added in a future release. 
> For now, manually create archives by building twice: once with the --use-xcframeworks option, and once without. 
> Then, create a zip from the Carthage/Build folder.
> ```
I don't fully understand what do they mean here. For our Carthage example project it was enough to run the next command and that's it.
```
carthage update --use-xcframeworks
```
It downloaded `openssl.xcframework` and created `themis.xcframework` with all the needed slices.

vixentael commented:
> I don't fully understand what do they mean here

Maybe they mean this https://github.com/Carthage/Carthage#archive-prebuilt-frameworks-into-zip-files

vixentael commented on .github/workflows/test-objc.yaml:
```diff
@@ -172,22 +180,17 @@ jobs:
-        run: carthage bootstrap
-      - name: Build Carthage projects - excluding arm64
         run: |
-          # ------ beginning of workaround
```
I like how we switch one workaround to _another_ workaround :D

julepka replied:
macOS BigSur on github actions already supports Carthage 0.38. It looks like the 10.15 should be updated in a couple of days... probably we can wait a little with merging and work on other edits and remove the workaround even in this PR :)

julepka commented:
I've put back the arm64 workaround for example projects. Also, replaced regular `themis.framework` in unit tests with `themis.xcframeworks`. Added `--use-xcframeworks` flag, so Carthage build themis as XCF

julepka commented:
Okay, regarding tests:
`Unit tests (CocoaPods)` - good
`Unit tests (Carthage)` - good
`Carthage project` - was good, became red for no reason (API rate limit, timeout, Carthage cache errors)
`CocoaPods project` - good  
`Code examples` - macOS Carthage failing due to API rate limit... Hmmm...

@ilammy @vixentael Does it make sense to break up our iOS/macOS tests into 2 parts because of constantly hitting the API rate limit? I don't understand why macOS only examples are having this problem.... hmm... Another thought is that we will be updating our tests to support the latest Themis version, and maybe the issue will be resolved on its own... Also, right now example projects may be mixing regular frameworks and XCF and it may be causing some issues as well. 

Okay. As for now, it looks like tests are passing well enough. I will continue with PR edits (per comments above).

vixentael commented:
> @ilammy @vixentael Does it make sense to break up our iOS/macOS tests into 2 parts because of constantly hitting the API rate limit?

yes, we can try to split those tests but in a separate PR

ilammy commented:
> Does it make sense to break up our iOS/macOS tests into 2 parts because of constantly hitting the API rate limit?

I'm not sure it would help. I mean, it's still the same machine pool that's going to run those tests.

The API rate limits are enforced by GitHub (because that's where Carthage pulls binaries from). Then it seems like it's trying to build Themis from source (as a fallback) and fails.

I thought that GitHub Action runners would get the necessary API keys in there for GitHub authentication. But I guess Carthage needs some help with that. There's https://github.com/Carthage/Carthage/pull/605 that implements it, and there's also https://github.com/Carthage/Carthage/issues/3058 which would make it automatic. I believe you just need to set `GITHUB_ACCESS_TOKEN` environment variable to `GITHUB_TOKEN` secret value:
```yaml
env:
  GITHUB_ACCESS_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

julepka commented:
@ilammy @vixentael 
Tests (almost) green, violets are blue
The PR is ready for another review :)

Rate limit issue - fixed
macOS example projects to be fixed with the update. I'm pretty sure that the remaining issue is that Carthage downloads XCF but the projects has a path to a regular framework... so it will be an easy fix.

vixentael commented on CHANGELOG.md:
```diff
@@ -4,33 +4,48 @@
 
 Changes that are currently in development and have not been released yet.
 
+## [0.13.9](https://github.com/cossacklabs/themis/releases/tag/0.13.9), May 14th 2021
```
👀

ilammy commented:
A poetry contest?
This I can certainly do.
But maybe not today.

vixentael commented on CHANGELOG.md:
```diff
@@ -4,33 +4,48 @@
+
+**Hotfix for Apple platforms:**
+
+- `themis` for Carthage switched to using XCFrameworks ([#817](https://github.com/cossacklabs/themis/pull/817)). So, the minimum required Carthage version is now [0.38.0](https://github.com/Carthage/Carthage/releases/tag/0.38.0).
```
```suggestion
- `themis` for Carthage switched to using XCFrameworks ([#817](https://github.com/cossacklabs/themis/pull/817)). So, the minimum required Carthage version is now [0.38.0](https://github.com/Carthage/Carthage/releases/tag/0.38.0). You can continue using previous Themis version with previous Carthage versions.
```

vixentael commented on CHANGELOG.md:
```diff
@@ -4,33 +4,48 @@
+
+- **Objective-C / Swift**
+
+  - `themis` for Carthage now pulls OpenSSL dependency as XCFramework, and Carthage builds `themis` as XCFramework as well. `Themis.xcodeproj` now uses `openssl.xcframwork` and `themis.xcframework`. Carthage dependencies should be built with `--use-xcframeworks` flag ([#817](https://github.com/cossacklabs/themis/pull/817)).
+  - Updated OpenSSL to the latest 1.1.1k for Carthage ([#817](https://github.com/cossacklabs/themis/pull/817)).
+  - Tests (Github Actions) are updated to use the latest Carthage version (0.38.0 and up) and `--use-xcframeworks` flag ([#817](https://github.com/cossacklabs/themis/pull/817)).
+
```
this should be separately documented in product docs as well

julepka commented:
Aaaaand Carthage on github machines updated today to 0.38.0 so I'll remove the workaround with the final-final edits )

ilammy commented on .github/workflows/test-objc.yaml:
```diff
@@ -119,11 +119,15 @@ jobs:
       - name: Check out code
         uses: actions/checkout@v2
       - name: Pull Carthage dependencies
-        run: carthage bootstrap
+        run: |
+          carthage bootstrap --use-xcframeworks
+          carthage build --no-skip-current --use-xcframeworks
```
Might be nice to split this into two steps, like in `project-carthage`, at least for consistency.
